Add starter and sample for Apache Geode.

jxblum opened the pull request:
Closes gh-5438.

jxblum commented:
DO NOT MERGE (yet)!

So, it turns out that the `spring-boot-starters/spring-boot-starter-data-geode` (and sample) I created have not effect.  Meaning, if you run the `SampleDataGeodeApplicationTest`, it picks up _Pivotal GemFire_ **8.2.0** and not _Apache Geode_ **1.0.0-incubating.M1** as it should.

I believe this is directly related to the [inclusion of the Spring Data Maven BOM file](https://github.com/spring-projects/spring-boot/blob/v1.4.0.M1/spring-boot-dependencies/pom.xml#L1855-L1861) based on the [Spring Data version](https://github.com/spring-projects/spring-boot/blob/v1.4.0.M1/spring-boot-dependencies/pom.xml#L135) currently used by _Spring Boot_.

The problem is 1) the _Spring Data_ Maven BOM file has a hard version on the `spring-data-gemfire` dependency in the dependency declarations, rather than a property placeholder (e.g. `${spring-data-gemfire.version}`) and a corresponding property definition...

```
<properties>
  <spring-data-gemfire.version>1.8.0.RC1</spring-data-gemfire.version>
<properties>
```

Thus as I have done in the `spring-boot/spring-boot-dependencies.pom` file, `properties` section, and therefore, making the Spring Data GemFire version included by Spring Boot overridable, if even possible.

I am thinking at this point, if there is not special _Maven_ magic to work around this issue, then we might have to go with an actual, separate _Spring Data GemFire_ dependency, as in **spring-data-geode**, making "_Spring Data GemFire with Apache Geode support_" a top-level module in the _Spring Data_ portfolio.

Thoughts?

snicoll commented on spring-boot-dependencies/pom.xml:
```diff
@@ -135,6 +135,7 @@
 		<spring-amqp.version>1.6.0.M1</spring-amqp.version>
 		<spring-cloud-connectors.version>1.2.1.RELEASE</spring-cloud-connectors.version>
 		<spring-batch.version>3.0.6.RELEASE</spring-batch.version>
+		<spring-data-geode.version>1.7.0.APACHE-GEODE-EA-M1</spring-data-geode.version>
```
Unrelated but that's a weird version naming. Any reason why it has to be that long?

jxblum replied:
As I explained [here](https://spring.io/blog/2015/06/12/spring-data-gemfire-supports-apache-geode) and [here](https://groups.google.com/a/pivotal.io/d/msg/spring-developer/KMbUfmlS_jA/F_EiftPFBwAJ), this is not a usual version number.

Essentially, 2 separate releases of _Spring Data GemFire_ are occurring off 1 codebase (in GitHub) right now.  Though, the _Spring Data GemFire Support for Apache Geode_ is technically occurring off a branch (`apache-geode`).

Based on the issues I had with the starter, I think now, even more than ever/before, it is time to consider _Spring Data GemFire with Apache Geode Support_ as a separate project (SD module) having a separately releasable artifact (i.e. `spring-data-geode`) and intuitive version number (i.e. `1.0.0.M1`).

snicoll commented on spring-boot-samples/spring-boot-sample-data-geode/.gitignore:
```diff
@@ -0,0 +1 @@
+apachegeode-maven-dependency-exclusions.txt
```
Should this have been committed?

jxblum replied:
No, it was/is for my own personal reference (i.e. not absolutely required) and technically [comes from the work](https://github.com/jxblum/spring-session/blob/apache-geode/spring-session-data-geode/build.gradle#L11-L19) on _Spring Session_ to include support for _Apache Geode_ through _Spring Data GemFire_, which has not been integrated yet.

Also, the _GemFire_ team, at my direction (see [GEODE-27](https://issues.apache.org/jira/browse/GEODE-27) for further details), has been busy cleaning up the mess that is _Apache Geode's_ dependencies, so these "exclusions" should not be necessary in the future.

The exclusions were only necessary in _Spring Session_ since they pull in and combine conflicting web related dependencies (e.g. `javax.servlet-api`), which was caused by _Apache Geode's_ unnecessary and incomplete dependency on _Apache HBase_, which pulls in Jetty, which pulls in javax.servlet, and so on, ugh.

snicoll commented on spring-boot-samples/spring-boot-sample-data-geode/pom.xml:
```diff
@@ -0,0 +1,57 @@
+	</organization>
+	<properties>
+		<main.basedir>${basedir}/../..</main.basedir>
+		<spring-shell.version>1.1.0.RELEASE</spring-shell.version>
```
A sample shouldn't override dependencies, it should just reuse the starter. If we don't, then it means that any user would have to do the same thing which defeats the purpose of the starter.

jxblum replied:
1. I could not find the _Spring Shell_ **version** declared anywhere in _Spring Boot_ (e.g. `spring-boot-dependencies/pom.xml`).

2. There is no `spring-boot-starter-shell`, or the like for _Spring Shell_ in `[spring-boot-starters](https://github.com/jxblum/spring-boot/tree/master/spring-boot-starters)` (AFAICT).

3. The `spring-boot-starter-data-gemfire/geode` starters do not declare the `spring-shell` dependency directly, since _Spring Shell_ is not a hard, required dependency.  It is only needed by GemFire/Geode if a Server is also to become a "Manager" (managing node in the cluster, which it is not be default and so needs to be explicitly enabled).  However, that does not prevent GemFire or Geode from erroneously trying to load the GemFire shell "command" classes, which are also used by the Manager within a cluster to carry out the remote, or cluster-based commands.  These command classes implement _Spring Shell_ classes/interfaces.  If they are not found, then GemFire/Geode throws a rather annoying (noisy), but benign Exception.  You can see this on GemFire/Geode startup if you remove the "explicit" "runtime-only" _Spring Shell_ dependency.

So, technically, the `spring-shell.version` is not been overridden (AFAIK).  It is also unnecessary but was put there to suppress the Exception stack trace in the log output, which is meaningless and confusing in this sample application context.

snicoll replied:
All dependencies in Spring Boot have to be declared in `spring-boot-dependencies`. A sample should be as straightforward as possible: if you have to define special entries or exclusion in the sample it means that everyone else should. This defeats the purpose of the starter.

If you want to showcase the "Manager" case in the sample, that's perfectly fine to add `spring-shell` explicitly (i.e. the starter should not bring _everything_ that is possible). But it shouldn't have a version there.

jxblum replied:
1. The exclusion of `spring-boot-starter-logging` from the `spring-boot-starter` dependency is to avoid all the verbose logging from _Spring Boot_.  I will remove it.

2. I am not showcasing a "Manager" in the sample.  I am simply avoiding the nasty stack trace spit out by GemFire/Geode on startup when it attempts to find and load the (_Spring Shell_) command classes, regardless of whether the GemFire node actually becomes a Manager or **not**.

The _Spring Shell_ command classes are **only** required by a managing node, but if the GemFire Server does not become a Manager, GemFire/Geode will try to find/load the command classes anyway, which is a bug in GemFire/Geode that confuses users all the time.

The _Spring Shell_ dependency definitely does **not** belong in the starter; it is not a requirement to start a GemFire Server, and whether a GemFire Server node is configured to become a Manager or not is completely optional and up to the user.  The sample is not.

Now, I can declare the _Spring Shell_ version in [spring-boot-dependencies/pom.xml](https://github.com/spring-projects/spring-boot/blob/master/spring-boot-dependencies/pom.xml#L45-L171), but...

1. Both _Pivotal GemFire_ and _Apache Geode_ require different versions to find/load the command classes appropriately.  _GemFire_ requires _Spring Shell_ 1.0.0.RELEASE.  _Geode_ (now) uses 1.1.0.RELEASE, a conundrum to be sure.

2. TMK, there is **no** support in _Spring Boot_ for _Spring Shell_ (e.g. `spring-boot-starter-shell`), so putting the _Spring Shell_ version in the [spring-boot-dependencies/pom.xml](https://github.com/spring-projects/spring-boot/blob/master/spring-boot-dependencies/pom.xml#L45-L171) seems fruitless.  Plus, I would still need to override the version used by Geode (or GemFire, since they are different, see #1).

The only other option is to let the nasty stack trace appear when users run the (GemFire/Geode) sample(s)...

```
[info 2016/03/22 16:44:50.450 PDT <main> tid=0x1] Command Service could not be initialized. Could not find Spring Shell library which is needed for CLI/gfsh in classpath. Internal support for CLI & gfsh is not enabled. Note: For convenience, absolute path of "gfsh-dependencies.jar" from "lib" directory of GemFire product distribution can be included in CLASSPATH of an application.
com.gemstone.gemfire.management.DependenciesNotFoundException: Could not find Spring Shell library which is needed for CLI/gfsh in classpath. Internal support for CLI & gfsh is not enabled. Note: For convenience, absolute path of "gfsh-dependencies.jar" from "lib" directory of GemFire product distribution can be included in CLASSPATH of an application.
	at com.gemstone.gemfire.management.cli.CommandService.createLocalCommandService(CommandService.java:135)
	at com.gemstone.gemfire.management.internal.beans.MemberMBeanBridge.<init>(MemberMBeanBridge.java:352)
	at com.gemstone.gemfire.management.internal.beans.ManagementAdapter.handleCacheCreation(ManagementAdapter.java:152)
	at com.gemstone.gemfire.management.internal.beans.ManagementListener.handleEvent(ManagementListener.java:111)
	at com.gemstone.gemfire.distributed.internal.InternalDistributedSystem.notifyResourceEventListeners(InternalDistributedSystem.java:2245)
	at com.gemstone.gemfire.distributed.internal.InternalDistributedSystem.handleResourceEvent(InternalDistributedSystem.java:505)
	at com.gemstone.gemfire.internal.cache.GemFireCacheImpl.init(GemFireCacheImpl.java:1009)
	at com.gemstone.gemfire.internal.cache.GemFireCacheImpl.create(GemFireCacheImpl.java:682)
	at com.gemstone.gemfire.cache.CacheFactory.create(CacheFactory.java:182)
	at com.gemstone.gemfire.cache.CacheFactory.create(CacheFactory.java:229)
	at org.springframework.data.gemfire.CacheFactoryBean.createCache(CacheFactoryBean.java:356)
	at org.springframework.data.gemfire.CacheFactoryBean.resolveCache(CacheFactoryBean.java:250)
	at org.springframework.data.gemfire.CacheFactoryBean.init(CacheFactoryBean.java:191)
	at org.springframework.data.gemfire.CacheFactoryBean.getObject(CacheFactoryBean.java:178)
	at org.springframework.data.gemfire.CacheFactoryBean.getObject(CacheFactoryBean.java:88)
	at org.springframework.beans.factory.support.FactoryBeanRegistrySupport.doGetObjectFromFactoryBean(FactoryBeanRegistrySupport.java:168)
	at org.springframework.beans.factory.support.FactoryBeanRegistrySupport.getObjectFromFactoryBean(FactoryBeanRegistrySupport.java:103)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getObjectForBeanInstance(AbstractBeanFactory.java:1590)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:317)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:197)
	at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:186)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.findAutowireCandidates(DefaultListableBeanFactory.java:1208)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1048)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1018)
	at org.springframework.beans.factory.support.ConstructorResolver.resolveAutowiredArgument(ConstructorResolver.java:834)
	at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:741)
	at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:467)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1123)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1018)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:510)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:482)
	at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:306)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:302)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:202)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.getBeansOfType(DefaultListableBeanFactory.java:535)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.getBeansOfType(DefaultListableBeanFactory.java:524)
	at org.springframework.context.support.AbstractApplicationContext.getBeansOfType(AbstractApplicationContext.java:1164)
	at org.springframework.data.gemfire.repository.support.GemfireRepositoryFactoryBean.setApplicationContext(GemfireRepositoryFactoryBean.java:71)
	at org.springframework.context.support.ApplicationContextAwareProcessor.invokeAwareInterfaces(ApplicationContextAwareProcessor.java:121)
	at org.springframework.context.support.ApplicationContextAwareProcessor.postProcessBeforeInitialization(ApplicationContextAwareProcessor.java:97)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.applyBeanPostProcessorsBeforeInitialization(AbstractAutowireCapableBeanFactory.java:408)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1570)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:545)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:482)
	at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:306)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:302)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:197)
	at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:186)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.findAutowireCandidates(DefaultListableBeanFactory.java:1208)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1048)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1018)
	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor$AutowiredFieldElement.inject(AutowiredAnnotationBeanPostProcessor.java:570)
	at org.springframework.beans.factory.annotation.InjectionMetadata.inject(InjectionMetadata.java:88)
	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor.postProcessPropertyValues(AutowiredAnnotationBeanPostProcessor.java:350)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.populateBean(AbstractAutowireCapableBeanFactory.java:1214)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:543)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:482)
	at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:306)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:302)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:197)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:775)
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:841)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:540)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:773)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:368)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:315)
	at org.springframework.boot.test.SpringApplicationContextLoader.loadContext(SpringApplicationContextLoader.java:98)
	at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContextInternal(DefaultCacheAwareContextLoaderDelegate.java:98)
	at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContext(DefaultCacheAwareContextLoaderDelegate.java:116)
	at org.springframework.test.context.support.DefaultTestContext.getApplicationContext(DefaultTestContext.java:83)
	at org.springframework.test.context.support.DependencyInjectionTestExecutionListener.injectDependencies(DependencyInjectionTestExecutionListener.java:117)
	at org.springframework.test.context.support.DependencyInjectionTestExecutionListener.prepareTestInstance(DependencyInjectionTestExecutionListener.java:83)
	at org.springframework.test.context.TestContextManager.prepareTestInstance(TestContextManager.java:230)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.createTest(SpringJUnit4ClassRunner.java:228)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner$1.runReflectiveCall(SpringJUnit4ClassRunner.java:287)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.methodBlock(SpringJUnit4ClassRunner.java:289)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:247)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:94)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.springframework.test.context.junit4.statements.RunBeforeTestClassCallbacks.evaluate(RunBeforeTestClassCallbacks.java:61)
	at org.springframework.test.context.junit4.statements.RunAfterTestClassCallbacks.evaluate(RunAfterTestClassCallbacks.java:70)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.run(SpringJUnit4ClassRunner.java:191)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:78)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:212)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:68)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:140)
```

I just think it does not look good for us that stack traces appear for no apparent (good) reason.

However, I am prepared and willing to do whatever in this case.

snicoll commented on spring-boot-samples/spring-boot-sample-data-geode/src/main/java/sample/data/geode/SampleDataGeodeApplication.java:
```diff
@@ -0,0 +1,161 @@
+	}
+
+	@Bean
+	Properties gemfireProperties(@Value("${sample.data.geode.log.level:config}") String logLevel) {
```
Let's not use `@Value` please. We have `@ConfigurationProperites` and meta-data support for showcasing custom key assistance in the IDE. Let me know if you need help to configure that.

jxblum replied:
I am not entirely sure I understand the meaning of your comment...

1. I did declare the `@ConfigurationProperties` annotation on my `SampleDataGeodeApplication` class, which is the source for these "custom" sample application properties when explicitly specified by the user on the command-line while running the sample (manually, or with the test), defaulting otherwise.

2. However, they do allow a user to do the following when running the sample from the command-line (not/or within the IDE)...

`$ mvn -pl spring-boot-sample-data-geode clean test -Dsample.data.geode.log.level:config=warn`

I need help understanding the intent as well as the proper way to configure what you are stating.  I also assume (hope) the "custom key assistance in the IDE" is IDE agnostic?  I did not specifically think IDE when I coded my sample as I often do things from the command-line (and only occasionally in my IDE (IntelliJ IDEA)).

snicoll replied:
Can you review the documentation maybe? 

The `@ConfigurationProperties` on the app is wrong. What are you expecting to happen exactly? It's definitely not what you wrote. The command line thing is unrelated: `@ConfigurationProperties` can be updated by the environment regardless of the source (command line, command line switch, application.properties, etc).

The bottom line: don't use `@Value`. You must have a POJO that represents the configuration keys that you want to expose in your sample. Of course that's IDE agnostic (this feature is here for quite some time...)

jxblum replied:
So, I think I completely misunderstood the meaning and purpose of `@ConfigurationProperties`.  My bad!

I assumed it was used to enable something akin to a `PropertyPlaceholderConfigurer` (e.g. `PropertySourcesPlaceholderConfigurer`) to resolve property placeholders/configuration (in `@Value` annotations used within application code) from some external source, as one possible feature among others.  I was basing my knowledge on [this](http://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#boot-features-external-config), to which now I definitely see I am incorrect.

Interestingly, my sample code does work as I expect and the property placeholder values (in the `@Value` annotations) are resolved when I specify them on the command-line (or my IDE) or not (thus using the default values).

I have never really tried [this](http://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#boot-features-external-config-typesafe-configuration-properties), but seems like a pretty slick way to manage configuration properties.

I will clean up my mess! ;-)

snicoll replied:
The reason why it works is because you have _nothing to do_ for `@Value` resolution to work ;) We are discouraging users to use `@Value` so we shouldn't use it ourselves. Thanks!

snicoll commented on spring-boot-samples/spring-boot-sample-data-geode/src/test/java/sample/data/geode/SampleDataGeodeApplicationTest.java:
```diff
@@ -0,0 +1,86 @@
+	@Autowired
+	private GemfireTemplate calculationsTemplate;
+
+	@Value("${sample.data.geode.calculator:factorial}")
```
ditto. You probably need a `GeodeSampleProperties` or something.

jxblum replied:
I prefer not to use a specific properties file(s).  GemFire/Geode also contains some magic for "discovering" properties files and if the user start mixing and matching application properties with GemFire/Geode System properties, they are sure to get themselves in trouble since GemFire/Geode pukes on any non-recognized property.

The idea behind using the "custom, sample" application-specific properties was to allow the user to override them if they choose (e.g. `sample.data.geode.log.level` to affect the log output in Geode (defaulting to "config"), or the `sample.data.geode.calculator` to affect the calculation performed (defaulting to "factorial")) using reasonable defaults if the user does not specify anything.  I.e. no extraneous configuration is necessary, but the sample can be adjusted if the user desires.

Perhaps, I am just not as familiar as I should be with all of _Spring Boot's_ (application) configuration techniques and there is possibly a better/recommended way to do what I did?

I am also happy to remove these custom, sample application-specific properties if so desired.  Just say the word.

snicoll commented on spring-boot-starters/spring-boot-starter-data-geode/pom.xml:
```diff
@@ -0,0 +1,60 @@
+		<dependency>
+			<groupId>org.springframework.data</groupId>
+			<artifactId>spring-data-gemfire</artifactId>
+			<version>${spring-data-geode.version}</version>
```
OK, I am confused here. You're overriding the version of `spring-data-gemfire` to use another version. That's not right IMO because we can't compose things anymore. We use the release train of spring data and adding an additional module is fine (I guess) but overriding an existing one is not.

jxblum replied:
Your confusion is understandable.

Currently, if a user wants to include _Spring_ support for _Apache Geode_, then the user only needs to switch the version of _Spring Data GemFire_ that she is using.

For instance, she might be using _Spring Data GemFire_ for _Pivotal GemFire_, at version **1.8.0.RC1** (based on _GemFire 8.2_).  Or, perhaps she is using a release version of SDG, **1.7.4.RELEASE** (based on _GemFire 8.1_).  Either way, she may later decide to kick the tires on _Apache Geode_.  If so, she only need switch the version of the _Spring Data GemFire_ dependency to **1.7.0.APACHE-GEODE-EA-M1** in her application POM file (perhaps with a profile).

In hindsight, this may or may not have been the best idea, but several factors played into the decision...

1. The `apache-geode` branch was supposed to be short-lived as _Pivotal GemFire_ and _Apache Geode_ converged.  However, since **April 2015** they continue to diverge and the disparity between the 2 codebase is only growing.  For instance, you cannot compile SDG with both _Pivotal GemFire_ and _Apache Geode_ interchangeably, hence the separate branch.

2.  The former was mostly due to the features in GemFire they decided withhold from Geode (e.g. CQs, WAN) to distinguish the open source and commercial offerings.  Now they are throwing everything into the open source domain (eventually).  But, even with that, the package structure is set to change in _Apache Geode_ from `com.gemstone.gemfire` to `org.apache.geode`.  At some point, when/if _Pivotal GemFire_ converges, it's package structure will need to change as well.  However, this is a significant undertaking and not without customer impact.  Clearly, both of these affect SDG right now.

3.  Even when/if _Pivotal GemFire_ and _Apache Geode_ converge, they are likely to always have a separate set of dependencies.

4.   At the time, _Apache Geode_ had no official release and so did not even have a Maven artifact.  Now that _Apache Geode_ `1.0.0-incubating.M1` has been released and is also available in Maven Central, it is easier to have fixed releases of SDG based on these _Apache Geode_ versions (M2 release is planned for March 25th).

5. etc...

Which leads me to (as I mentioned previously in my initial comment)...

It maybe necessary now to have a separate SD project module for _Spring Data GemFire_ with _Apache Geode_ support.  I.e. let's create a **Spring Data Geode** project/module (which does NOT currently exist).

This new SD module may or may not be included in the SD release train.  Technically, it actually makes more sense if the **new** _Spring Data Geode_ module releases with _Apache Geode_, at least until Geode reaches GA.  FTR, SD release trains are not tied to either _Pivotal GemFire_ or _Apache Geode_ releases (as with any other SD driver, e.g. MongoDB, for that matter), nor should it.

Regardless, if SDG with _Apache Geode_ support were a separate SD module (i.e. **Spring Data Geode**), then it would not be necessary to override the version of SDG brought in by the SD BOM POM.

However, my thinking here was that, technically, a user can override any dependency version specified by a particular _Spring Boot_ version (e.g. `1.3.3.RELEASE`) containing the "blessed", "curated" set of dependencies as defined in the `[spring-boot-depedencies/pom.xml](https://github.com/spring-projects/spring-boot/blob/master/spring-boot-dependencies/pom.xml#L45-L171)` file simply by **overriding** the desired dependency's version property in the user's application POM file, like so...

NOTE: by way of example, I will pick on "_GemFire_", which is [at **8.1.0** in _Spring Boot_ **1.3.3.RELEASE**](https://github.com/spring-projects/spring-boot/blob/v1.3.3.RELEASE/spring-boot-dependencies/pom.xml#L71))...

```
// application Maven pom.xml
<project ...>
  <modelVersion>4.0.0</modelVersion>
  <parent>
    <groupId>org.springframework.boot</groupId>
    <artifactId>spring-boot-starter-parent</artifactId>
    <version>1.3.3.RELEASE</version>
  </parent>
  <groupId>com.example</groupId>
  <artifactId>example-spring-boot-gemfire-app</artifactId>
...
  <properties>
    ...
    <gemfire.version>8.2.0</gemfire.version>
  </properties>
...
</project>
```

Even though _Spring Boot_ **1.3.3.RELEASE** is based on _GemFire_ **8.1**, a user can effectively "override" that version and use, say, _GemFire_ **8.2**, simply by redeclaring the "`<gemfire.version>`" property in her application POM file.  This is true for nearly all dependencies (except for the SD BOM POM; explanation below).

So, following this thinking, my idea was that the _Spring Boot_ "Starter" for _Apache Geode_ could provide this "override" automatically for the user and include the right version of _Spring Data GemFire_ that supports _Apache Geode_ for development purposes simply by including the `spring-boot-starter-data-geode`.  Of course, this "override" would be approved/blessed by us for the user's convenience.

Except for the fact (and the "explanation") the SD BOM POM does not use dependency properties (e.g. `<spring-data-gemfire.version>`) when declaring the SD module versions in a particular BOM/release train (e.g. Hopper), and therefore, the SDG version is not separately overridable, as may be desired by the user.

So, how is a user to experiment with say another version of SD GemFire, or Redis, or Mongo, and so on, perhaps an experimental version containing a particular feature, in the context of _Spring Boot_ until the version of that SD module is included in the SD release train (BOM POM) and that release train version has been updated by the user manually ([here](https://github.com/spring-projects/spring-boot/blob/v1.4.0.M1/spring-boot-dependencies/pom.xml#L135)), or by us in a subsequent _Spring Boot_ release.  Make sense?

As one example, consider @christophstrobl  built SD _Repository_ support for SD Redis on a separate branch (e.g. `issues/DATAREDIS-425`), which had build snapshots ([1.7.0.DATAREDIS-425-SNAPSHOT](https://repo.spring.io/webapp/#/artifacts/browse/tree/General/libs-snapshot-local/org/springframework/data/spring-data-redis/1.7.0.DATAREDIS-425-SNAPSHOT)).

A user would be able to try this feature out (which was not technically [released until Hopper-RC1, or rather SD Redis 1.7.0.RC1](http://docs.spring.io/spring-data/redis/docs/1.7.0.RC1/changelog.txt)) in _Spring Boot_ **1.4.0.M1**, or event **1.3.3.RELEASE** (which was well **before** the `Hopper-RC1` release), simply by just overriding, say a "`<spring-data-redis.version>`" property, had it existed.

Anyway, I think I am tempted to start the `spring-data-geode` discussion now with the SD team.

Let me know if you have additional thoughts or concerns.  I mainly wanted to share my thinking here, and I am also prepared to do whatever the _Spring Boot_ team feels is most appropriate.

Thanks.

snicoll replied:
My general comment is that Spring Boot can't have two conflicting opinion about the same coordinates. You can override whatever you want in your own app (and at your own risk). So if you don't like the version of log4j2 that we provide, you can choose a different one.

But we can't have one module in Spring Boot declares that we should use {{com.example;acme:1.0.0}} and another that we should use {{com.example.acme:2.0.0-incubating-m1}}. 

That's already what I said in a previous comment: all public dependencies _must_ be declared in `spring-boot-dependencies` and no version override is allowed in the Spring Boot codebase.

I think we should put this PR on hold until you sort this out.

jxblum replied:
The only option then is to create a new, top-level SD project by pulling the `apache-geode` branch out into a new source code repository in order to release a new SD module (i.e. **Spring Data Geode**) on which the _Spring Boot Starter_ will be based

A long time will pass before there exists a version of _Pivotal GemFire_ based on _Apache Geode_ where potentially SDG can support both _GemFire_ and _Geode_ simultaneously in a single release (if ever).

Without this, there is simply no way to add early access (preview-style) support for _Apache Geode_ in _Spring Boot_.

jxblum replied:
Also, _Spring Boot_ is currently limited in that a user (at her own risk) using a starter, or inheriting from the `spring-boot-starter-parent`, **cannot** override any particular _Spring Data_ module (e.g. _SD GemFire_, _SD Redis_, or otherwise) in a given _Spring Boot_ release (e.g. `1.3.3.RELEASE`) at all.  For instance...

// application pom.xml

<parent>
  <groupId>org.springframework.boot</groupId>
  <artifactId>spring-boot-starter-parent</artifactId>
  <version>1.4.0.BUILD-SNAPSHOT</version>
</parent>
...
<properties>
  <spring-data-gemfire.version>1.7.4.RELEASE</spring-data-gemfire.version>
  <spring-data-redis.version>1.6.4.RELEASE</spring-data-redis.version>
</properties>

<dependencies>
  <dependency>
    <groupId>org.springframework.boot</groupId>
    <artifactId>spring-boot-starter-data-gemfire</artifactId>
  </dependency>
  <dependency>
    <groupId>org.springframework.boot</groupId>
    <artifactId>spring-boot-starter-data-redis</artifactId>
  </dependency>
</dependencies>

Essentially, there is no reason to use the _Spring Boot_ Starters for either _GemFire_ or _Redis_ at this point since there is no way for the user to override the dependency version.  The user is effectively stuck with the `Hopper` versions, where as other [dependencies](https://github.com/spring-projects/spring-boot/blob/v1.4.0.M1/spring-boot-dependencies/pom.xml#L131-L151) (including _Spring_, and even _Spring_ core) can be overridden, and even other individual modules within other_Spring_ portfolio projects (e.g [Spring Security](https://github.com/spring-projects/spring-boot/blob/v1.4.0.M1/spring-boot-dependencies/pom.xml#L143-L145), or [Spring Social](https://github.com/spring-projects/spring-boot/blob/v1.4.0.M1/spring-boot-dependencies/pom.xml#L147-L150)).

Of course, these Spring projects may not include BOM for their collective modules.  But, what is similar is that not all SD modules in the SD portfolio (at a given version, e.g. [Hopper](https://github.com/spring-projects/spring-data-commons/wiki/Release-Train-Hopper#participating-modules)) are at the same version, individually.

snicoll replied:
>  cannot override any particular Spring Data module 

That's inaccurate. We are relying on the release train so you could switch that if you wanted to. You could also override the dependency for a particular module by specifying a dependency management for it (your example uses properties that do not exist but nothing prevents you to redeclare a dependency with a different version).

Having said that, that's why this PR is on hold. If you have to do that kind of things then it shouldn't be in Spring Boot proper.

jxblum replied:
> You could also override the dependency for a particular module by specifying a dependency management for it

Right, good point!

I could see a user wanting to do this over overriding the entire SD release train, presumably because the user wants specific (perhaps, upcoming) features in particular SD module, and is either not using the other SD modules, or happy with the SB recommended/vetted version.

snicoll commented on spring-boot-samples/spring-boot-sample-data-geode/src/main/java/sample/data/geode/SampleDataGeodeApplication.java:
```diff
@@ -0,0 +1,161 @@
+ * @since 1.4.0
+ */
+@SpringBootApplication
+@ConfigurationProperties
```
That's wrong. This should be`@EnableConfigurationProperties(SampleDataGeodeProperties.class)` and you should define that class with the list of keys that the sample needs.

jxblum replied:
Will do!

jxblum commented:
Note, I cleaned up both #5444 and this PR by...

1. Removing all uses of `@Value` annotations.
2. Along with removing all `<exclusions>` in the (sample) POMs.
3. I also moved the `spring-shell.version` property from the GemFire/Geode sample POMs to `spring-boot-dependencies/pom.xml`.  The GemFire/Geode starter POMs do **not** include _Spring Shell_ as a dependency because it is not required at runtime to use GemFire/Geode (it just avoids the nasty, benign stack trace in the samples).

#5444 (polishing the GemFire starter/sample) is complete.

This PR still needs to sort out the `spring-data-gemfire` artifact.  Again, I think the only appropriate solution is to create a new SD module, **Spring Data Geode**.

@olivergierke what say you?

jxblum commented:
Note, out of curiosity, I also observed that including the following in the `spring-boot-starter-data-geode` POM file...

```
<dependencyManagement>
	<dependencies>
		<dependency>
			<groupId>org.springframework.data</groupId>
			<artifactId>spring-data-gemfire</artifactId>
			<version>${spring-data-geode.version}</version>
		</dependency>
	</dependencies>
</dependencyManagement>
```

Rather than...

```
<dependency>
	<groupId>org.springframework.data</groupId>
	<artifactId>spring-data-gemfire</artifactId>
	<version>${spring-data-geode.version}</version>
```

With the explicit `<version>` is no different.  I.e. it has the same effect.

_Pivotal GemFire_ is still used over _Apache Geode_ in the `spring-boot-sample-data-geode` application, that just declares the `spring-boot-starter-data-geode` dependency.  This means the SD release train dependency declaration is ever lasting/overriding and takes absolute precedence.

The only workaround is to put the above `<dependencyManagement>` section in the Geode sample, which means the user would need to declare the same `<dependencyManagement>` section in their application POM, which defeats the purpose of the `spring-data-starter-data-geode` Starter since the user could just use the `spring-boot-starter-data-gemfire` Starter and include the same `<dependencyManagement>` section in her application POM to get _Apache Geode_ support.

Therefore, everything rides on creating a separate artifact (i.e. **Spring Data Geode**), or simply not having a _Spring Boot Starter_ for _Apache Geode_.  However, I do think it would be nice and convenient for our users.

**DISCLAIMER: I know that we don't want to include `<dependencyManagement>` section in either the Geode starter or sample POMs, as it is the same thing as overriding the version, which we DON'T want to do, I know; I was simply experimenting out of curiosity, ;-)**

jxblum commented:
@SpringBootTeam - Alright, after all checks pass, this PR is ready to go (_review_ and _merge_ after your approval).

If there are any questions or concerns, please let me know.

Thanks!
-John

wilkinsona commented:
thanks, @jxblum. What's the plan for a final release of both Geode and the accompanying Spring Data module? We can't use a milestone dependency in a final release of Spring Boot so this may have to wait till after 1.4.0.

jxblum commented:
Hi @wilkinsona -

There has only been [recent mention](http://markmail.org/search/?q=list%3Aorg.apache.geode.dev+order%3Adate-backward#query:list%3Aorg.apache.geode.dev%20order%3Adate-backward%20from%3A%22Dan%20Smith%22+page:1+mid:yuitdtajrd4qdezd+state:results) of an official _Apache Geode_ **1.0.0 GA** release.  However, there is no current plan or schedule in place as of yet.  

Still, both _Apache Geode_ 1.0.0-incubating.M2 and _Spring Data Geode_ are [available in Maven Central](http://search.maven.org/#search%7Cga%7C1%7Cgeode).  Would it be possible to include Geode support as a sort of Early Access (EA) preview, citing that _Apache Geode_ has not reached final GA yet?

Technically, the only reason why I have Spring Data Geode tagged as a milestone (i.e. `APACHE-GEODE-INCUBATING-M2`) is because Geode is still releasing milestones.  However, SD Geode is really based on the latest `Spring Data GemFire` codebase (i.e. `1.8.1.RELEASE`) and is therefore, quite stable.

Cheers,
John

wilkinsona commented:
Thanks. We'll wait for Geode to reach GA.

jxblum commented:
@wilkinsona, @snicoll & Spring Boot team-

This PR to add support for Apache Geode has been updated with the latest release of _Spring Data Geode_ `1.0.0.INCUBATING-RELEASE` ([here](http://search.maven.org/#search%7Cga%7C1%7Cspring-data-geode)), which is **now** based on _Apache Geode_ `1.0.0-incubating` GA release ([here](http://search.maven.org/#search%7Cga%7C1%7Cgeode); details [here](http://geode.incubator.apache.org/releases/)).

If you have questions, let me know.

Thanks!

philwebb commented:
As with #6952 we should wait for a non incubating release I think.

wilkinsona commented:
As discussed, this code is going to be delivered in a separate module outside of Spring Boot.

jxblum commented:
@philwebb, @wilkinsona - I heard the message loud and clear on the SD Geode _Caching_ and _Repository_ support, but why does the _Spring Boot_ "starter" for _Spring Data Geode_ need to live outside of _Spring Boot_?  If that is the case, then why do we include the [spring-boot-starter-data-gemfire](https://github.com/spring-projects/spring-boot/tree/master/spring-boot-starters/spring-boot-starter-data-gemfire) in _Spring Boot_?  What's the difference?

wilkinsona commented:
The rule that we apply is that we only provide starters and samples for something if we also have auto-configuration for it. We've rejected other requests for starters and samples based on this rule and we need to apply it consistently. The GemFire starter is an exception to that rule for reasons that escape me. If we were doing things again, we wouldn't have that starter either.
